File: problems/1792_maximum_average_pass_ratio/1792_maximum_average_pass_ratio.java
class Solution {
    public double maxAverageRatio(int[][] classes, int extraStudents) {
        PriorityQueue<ClassRecord> pq = new PriorityQueue<>((a, b) -> {
            if (a.gain < b.gain)
                return 1;
            else if (a.gain > b.gain)
                return -1;
            else
                return 0;
        });

        for (int[] c : classes) {
            pq.add(new ClassRecord(c));
        }
        while (extraStudents > 0) {
            pq.add(pq.poll().addStudent());
            extraStudents--;
        }
        double sum = 0;
        while (!pq.isEmpty()) {
            ClassRecord cr = pq.poll();
            sum += (double) cr.pass / cr.total;
        }
        return sum / classes.length;
    }

    class ClassRecord {
        int pass;
        int total;
        double gain;

        public ClassRecord(int[] arr) {
            pass = arr[0];
            total = arr[1];
            gain = calculateGain();
        }

        public ClassRecord addStudent() {
            pass++;
            total++;
            gain = calculateGain();
            return this;
        }

        public double calculateGain() {
            return (pass + 1.0) / (total + 1.0) - (double) pass / total;
        }
    }
}

File: problems/1792_maximum_average_pass_ratio/1792_maximum_average_pass_ratio.py
class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        def extra_gain(p, t):
            return (p + 1) / (t + 1) - p / t
        
        heap = [(-extra_gain(p, t), p, t) for p, t in classes]
        heapify(heap)

        for _ in range(extraStudents):
            neg_gain, p, t = heappop(heap)
            p, t = p + 1, t + 1
            heappush(heap, (-extra_gain(p, t), p, t))        
        
        return sum(p / t for _, p, t in heap) / len(classes)